include/led.h

```diff
@@ -4,20 +4,109 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef _ZEPHYR_LED_STRIP_H
-#define _ZEPHYR_LED_STRIP_H
+#ifndef _ZEPHYR_LED_H
+#define _ZEPHYR_LED_H
 
 /**
  * @file
- * @brief Public API for controlling linear strips of LEDs.
- *
- * This library abstracts the chipset drivers for individually
- * addressable strips of LEDs.
+ * @brief Public API for controlling LEDs
  */
 
 #include <zephyr/types.h>
 #include <device.h>
 
+/**
+ * @typedef led_api_update_duty
+ * @brief Callback API for updating an LED duty cycle
+ *
+ * @see led_update_duty() for argument descriptions.
+ */
+typedef int (*led_api_update_duty)(struct device *dev, int id, int duty);
+
+/**
+ * @typedef led_api_update_freq
+ * @brief Callback API for updating an LED frequency
+ *
+ * @see led_update_freq() for argument descriptions.
+ */
+typedef int (*led_api_update_freq)(struct device *dev, int id, int freq);
+
+/**
+ * @typedef led_api_update_channel
+ * @brief Callback API for updating an LED channel
+ *
+ * @see led_update_channel() for argument descriptions.
+ */
+typedef int (*led_api_update_channel)(struct device *dev, int channel, int mode);
+
+/**
+ * @brief LED driver API
+ *
+ * This is the mandatory API that any LED driver needs to expose.
+ */
+struct led_driver_api {
+	led_api_update_duty update_duty;
+	led_api_update_freq update_freq;
+	led_api_update_channel update_channel;
+};
+
+/**
+ * @brief Update an LED device duty cycle
+ *
+ * This routine updates the LED device duty cycle
+ *
+ * @param dev LED device
+ * @param id ID of the LED duty cycle register
+ * @param duty Duty cycle value
+ * @return 0 on success, negative on error
+ */
+static inline int led_update_duty(struct device *dev,
+				    int id,
+				    int duty)
+{
+	const struct led_driver_api *api = dev->driver_api;
+
+	return api->update_duty(dev, id, duty);
+}
+
+/**
+ * @brief Update an LED device frequency
+ *
+ * This routine updates the LED device frequency
+ *
+ * @param dev LED device
+ * @param id ID of the LED frequency register
+ * @param freq Frequency value
+ * @return 0 on success, negative on error
+ */
+static inline int led_update_freq(struct device *dev,
+				  int id,
+				  int freq)
+{
+	const struct led_driver_api *api = dev->driver_api;
+
+	return api->update_freq(dev, id, freq);
+}
+
+/**
+ * @brief Update an LED device channel
+ *
+ * This routine updates the LED device channel
```

Comment:
I don't know what a "channel" means in this context. In the LED strip APIs, this is defined as either an individual RGB color channel in an addressable RGB LED (as in the led_strip_update_rgb() documentation), or otherwise one of the individually addressable LEDs that appears linearly in the strip (as in led_strip_update_channels()).

Do you mean the same thing here? If so, can you add some documentation saying so? Also, if that's what you mean, did you leave out RGB support for some reason? I can imagine people wanting to set RGB values explicitly for an individual LED (think of the LED HAL APIs on an Android phone that lets apps set a specific color in given conditions, so you can make emails set the LED red, IRC messages set it blue, etc.), so RGB support would be good to have as well.

Either way, can you try to define more clearly what is meant here? Otherwise, driver writers will get confused.

Reply:
Channel refers to individual LEDs here, will add more documentation. Also, I was thinking about adding RGB support later but anyhow it doesn't harm to do it now :)